dex/meter/meterer.go

```diff
@@ -0,0 +1,72 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package meter
+
+import (
+	"sync"
+	"time"
+)
+
+// SignalMeterer meters a simple signal to prevent tight grouping. The user
+// specifies a minimum delay (minDelay). If a signal arrives within minDelay of
+// the previous signal, that signal will be scheduled for later to respect the
+// configured delay. If multiple signals arrive within minDelay, they will be
+// grouped into a single delayed signal. Error signals are never delayed.
+type SignalMeterer struct {
+	minDelay time.Duration
+	recv     func(error)
+
+	mtx       sync.Mutex
+	last      time.Time
+	scheduled *time.Timer
+}
+
+// NewSignalMeterer creates a new SignalMeterer.
+func NewSignalMeterer(minDelay time.Duration, recv func(error)) *SignalMeterer {
+	return &SignalMeterer{
+		minDelay: minDelay,
+		recv:     recv,
+	}
+}
+
+// Signal sends a signal, with a delay if appropriate.
+func (s *SignalMeterer) Signal(err error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	// Errors are passed immediately.
+	if err != nil {
+		go s.recv(err)
+		return
+	}
+	// If a signal is scheduled, nothing to do.
+	if s.scheduled != nil {
+		return
+	}
+	if time.Since(s.last) > s.minDelay {
+		// It's been long enough. No metering required.
+		go s.recv(nil)
+		s.last = time.Now()
+		return
+	}
+	// Meter the signal.
+	delay := time.Until(s.last.Add(s.minDelay))
+	s.scheduled = time.AfterFunc(delay, func() {
+		s.mtx.Lock()
+		s.scheduled = nil
+		s.last = time.Now()
+		s.mtx.Unlock()
+		s.recv(nil)
+	})
+}
+
+// Stop cancels any scheduled signals.
+func (s *SignalMeterer) Stop() {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	if s.scheduled != nil {
+		if !s.scheduled.Stop() {
+			<-s.scheduled.C
+		}
+	}
```

Comment:
maybe also need to `s.scheduled = nil` inside of this `if` (assuming `Stop` method is only meant to `cancels any scheduled signals` and not completely prevent further usage of `SignalMeterer`),

as the following condition will be hit, it seems:

<img width="715" alt="image" src="https://user-images.githubusercontent.com/7031569/170839611-cd2d9ae9-ed28-4041-9883-1b25ac165743.png">

Reply:
> and not completely prevent further usage of `SignalMeterer`

I always just assumed this was a one-`Done` type.